build-system/tasks/e2e/controller-promise.js

```diff
@@ -80,14 +79,18 @@ class ControllerPromise {
  * Wrap the given wait function with the given mutation function,
  * while still allowing it to be mutated again in the future by
  * the inner opt_mutate function.
- * @param {function(TYPE,function(TYPE): ?TYPE): !Promise=} wait
- * @param {function(TYPE): TYPE} mutate
- * @return {!Promise}
- * @template TYPE
+ * @param {function(CONDITION, function(VALUE): ?DERIVED): Promise<RES>} wait
+ * @param {function(VALUE): MUTANT} mutate
+ * @return {function(CONDITION, function(MUTANT): ?DERIVED): Promise<RES>}
+ * @template CONDITION
+ * @template MUTANT
+ * @template DERIVED
+ * @template VALUE
+ * @template RES
  */
 function wrapWait(wait, mutate) {
   return (condition, opt_mutate) => {
-    opt_mutate = opt_mutate || ((x) => x);
+    opt_mutate = opt_mutate || ((x) => /** @type {*} */ (x));
     return wait(condition, (value) => opt_mutate(mutate(value)));
   };
 }
```

Comment:
This was a heck of a puzzle. More than open to changing type names etc...

Reply:
Go for it!